Closes #73: add ADVS example and update ADSL example and corresponding specs (#83)
pharmaverse/examples: adam/adsl.qmd

library(stringr)

# Read in input SDTM data
dm <- pharmaversesdtm::dm
ds <- pharmaversesdtm::ds
ex <- pharmaversesdtm::ex
ae <- pharmaversesdtm::ae
vs <- pharmaversesdtm::vs
suppdm <- pharmaversesdtm::suppdm

# When SAS datasets are imported into R using haven::read_sas(), missing
# character values from SAS appear as "" characters in R, instead of appearing
# as NA values. Further details can be obtained via the following link:
# https://pharmaverse.github.io/admiral/articles/admiral.html#handling-of-missing-values
dm <- convert_blanks_to_na(dm)
ds <- convert_blanks_to_na(ds)
ex <- convert_blanks_to_na(ex)
ae <- convert_blanks_to_na(ae)
vs <- convert_blanks_to_na(vs)
suppdm <- convert_blanks_to_na(suppdm)

# Combine Parent and Supp - very handy! ----
dm_suppdm <- combine_supp(dm, suppdm)
```

Next we need to load the specification file in the form of a `{metacore}` object.

```{r metacore, warning=FALSE, results='hold'}
# Read in metacore object
metacore <- spec_to_metacore(
  path = "./metadata/ADSL_ADVS_spec.xlsx",
  where_sep_sheet = FALSE,

>   quiet = TRUE

Reviewer: Could we mention why we are setting these arguments like this?

Author: `where_sep_sheet` explained
`quiet` dropped and thus left to the default value since we do not have any warnings to mute

) %>%
  select_dataset("ADSL")
```
## Start Building Derivations

The first derivation step we are going to do is to pull through all the columns that come directly from the SDTM datasets.

In this case all the columns come from `DM` and `SUPPDM`so that are the only datasets we will pass into `metatools::build_from_derived()`.
The resulting dataset has all the columns combined and any columns that needed renaming between SDTM and ADaM are renamed.

```{r demographics}
adsl_preds <- build_from_derived(metacore,
                                 ds_list = list("dm" = dm_suppdm, "suppdm" = dm_suppdm),
                                 predecessor_only = FALSE, keep = FALSE)

> head(adsl_preds, n=10)

Reviewer: could we display this in a prettier way using reactable package?  https://github.com/pharmaverse/admiraldiscovery/blob/v0.3.0/R/interactive.R here is what Daniel uses it in admiraldiscovery

Author: to compare with `head()`
![image](https://github.com/user-attachments/assets/92d3abca-46e0-414b-86bf-d1db87dc8070)

Reviewer: we could just create another folder called `functions` at the top-level.  and store the fancy code in there.  and then your code could call the reactable code from the functions folder

Reviewer: but yeah agree the vanilla reactable looks sad

```

### Grouping variables

Now we have the base dataset, we can start to create some variables.
There are a few options to create grouping variables and their corresponding numeric variables.

> Option 1: We can start with creating the subgroups using the controlled terminology, in this case `AGEGR1`.

Reviewer: Could we make use of how the logs section does this?  This is a pretty long section 
![image](https://github.com/user-attachments/assets/52a49283-a2df-40ac-8510-c695586e46e4)

Author: @bms63 Could you let me know what you expect here using _`logr`_ / _`logrx`_ ? Thanks

Reviewer: `metatools option` | `admiral option`

maybe??

Author: As agreed with @bms63 today, it will be done later since I am lost with this request.

The metacore object holds all the metadata needed to make `ADSL`. Part of that metadata is the controlled terminology,
which can help automate the creation of subgroups. We can look into the `{metacore}` object and see the controlled terminology for `AGEGR1`.

```{r}
get_control_term(metacore, variable = AGEGR1)
```

Because this controlled terminology is written in a fairly standard format we can automate the creation of `AGEGR1`.
The function `metatools::create_cat_var()` takes in a `{metacore}` object, a reference variable -
in this case `AGE` because that is the continuous variable `AGEGR1` is created from, and the name of the sub-grouped variable.
It will take the controlled terminology from the sub-grouped variable and group the reference variables accordingly.

```{r grouping_option_1}
adsl_ct <- adsl_preds %>%
   create_cat_var(metacore, ref_var = AGE,
                  grp_var = AGEGR1, num_grp_var = AGEGR1N)

head(adsl_ct, n=10)
```

> Option 2: We can create the subgroups using the admiral::derive_vars_cat() function available from {admiral} v1.1.2. 

Reviewer: ```suggestion
Option 2: We can create the subgroups using the admiral::derive_vars_cat() function available since {admiral} v1.2.0. 
```

```{r codelist}
adsl_ct <- adsl_ct %>%
   create_var_from_codelist(metacore = metacore,
                            input_var = RACE,
                            out_var = RACEN)

head(adsl_ct, n=10)
```

### Exposure derivations

Now we have sorted out what we can easily do with controlled terminology it is time to start deriving some variables.
Here you could refer directly to using the `{admiral}` template and [vignette](https://pharmaverse.github.io/admiral/articles/adsl.html) in practice,
but for the purpose of this end-to-end ADaM vignette we will share a few exposure derivations from there.
We derive the start and end of treatment (which requires dates to first be converted from DTC to DTM), the treatment start time, the treatment duration, and the safety population flag.
Note that the populations flags are mainly company/study specific no dedicated functions are provided, but in most cases they can easily be derived using `admiral::derive_var_merged_exist_flag()`.

```{r exposure}

> ex_ext <- ex %>%

Reviewer: is this needed can we just use `ex` in the function

Author: Do you mean without imputing the Exposure start and end dates or just naming it `ex` instead of `ex_ext`?

# https://pharmaverse.github.io/admiral/articles/admiral.html#handling-of-missing-values
dm <- convert_blanks_to_na(dm)
ds <- convert_blanks_to_na(ds)
ex <- convert_blanks_to_na(ex)
ae <- convert_blanks_to_na(ae)
vs <- convert_blanks_to_na(vs)
suppdm <- convert_blanks_to_na(suppdm)

# Combine Parent and Supp - very handy! ----
dm_suppdm <- combine_supp(dm, suppdm)
```

Next we need to load the specification file in the form of a `{metacore}` object.

```{r metacore, warning=FALSE, results='hold'}
# Read in metacore object
metacore <- spec_to_metacore(
  path = "./metadata/ADSL_ADVS_spec.xlsx",
  # All datasets are described in the same sheet
  where_sep_sheet = FALSE
) %>%
  select_dataset("ADSL")
```
## Start Building Derivations

The first derivation step we are going to do is to pull through all the columns that come directly from the SDTM datasets.

In this case all the columns come from `DM` and `SUPPDM`so that are the only datasets we will pass into `metatools::build_from_derived()`.
The resulting dataset has all the columns combined and any columns that needed renaming between SDTM and ADaM are renamed.

```{r demographics}
adsl_preds <- build_from_derived(metacore,

>                                  ds_list = list("dm" = dm_suppdm, "suppdm" = dm_suppdm),

Author: We combined `[DM]` and `[SUPPDM]` at the beginning with `combine_supp()` to show case this fantastic function.
In the specs it is specified from which domain the variables come from:
![image](https://github.com/user-attachments/assets/7956dbe2-2842-472e-853f-840c0cfb7c8c)
We then get the `metacore` object with the corresponding domain to look for. Since we combined `[DM]` and `[SUPPDM]`, we need to specify that `[DM]` variables come from the `dm_suppdm` dataset that we just created, and `[SUPPDM]` variables also come from `dm_suppdm`.
![image](https://github.com/user-attachments/assets/26420812-4a01-4834-8762-0eb7b90150cd)
Setting `ds_list = list("dm" = dm_suppdm)` only does not gather the variables from `[SUPPDM]`
![image](https://github.com/user-attachments/assets/39cb3c93-87e9-486f-a29f-64cedacb4ab3)
If we want to separate `ds_list = list("dm" = dm, "suppdm" = suppdm)` we would need to transpose `[SUPPDM]` to get the parameters (`QNAM`) as variables.
![image](https://github.com/user-attachments/assets/16e67747-9131-41f3-b175-611af718b3bc)

>                                  ds_list = list("dm" = dm_suppdm, "suppdm" = dm_suppdm),

Author: will try to think how to shorter the explanation without screenshots 🤓

>                                  ds_list = list("dm" = dm_suppdm, "suppdm" = dm_suppdm),

Author: Hopefully clarified. Please let me knw if it could be improved

>                                  ds_list = list("dm" = dm_suppdm, "suppdm" = dm_suppdm),

Author: I am not sure about the meaning of the breaking change:
![image](https://github.com/user-attachments/assets/3fad186c-ff3a-4f38-b405-6b07f23cab0c)
Does it mean that it should run by calling only 1 reference, instead of both `"dm"` and `"suppdm"` as above?

>                                  ds_list = list("dm" = dm_suppdm, "suppdm" = dm_suppdm),

Reviewer 2: > Does it mean that it should run by calling only 1 reference, instead of both `"dm"` and `"suppdm"` as above?

I'm not sure. Did you try it out?
Maybe it means that you can specify `ds_list = list("dm" = dm_suppdm)` although there are variables of the from `suppdm.var` in the specs?

There are a few options to create grouping variables and their corresponding numeric variables.

Option 1: We can start with creating the subgroups using the controlled terminology, in this case `AGEGR1`.
The metacore object holds all the metadata needed to make `ADSL`. Part of that metadata is the controlled terminology,
which can help automate the creation of subgroups. We can look into the `{metacore}` object and see the controlled terminology for `AGEGR1`.

```{r}
get_control_term(metacore, variable = AGEGR1)
```

Because this controlled terminology is written in a fairly standard format we can automate the creation of `AGEGR1`.
The function `metatools::create_cat_var()` takes in a `{metacore}` object, a reference variable -
in this case `AGE` because that is the continuous variable `AGEGR1` is created from, and the name of the sub-grouped variable.
It will take the controlled terminology from the sub-grouped variable and group the reference variables accordingly.

```{r grouping_option_1}
adsl_ct <- adsl_preds %>%
   create_cat_var(metacore, ref_var = AGE,
                  grp_var = AGEGR1, num_grp_var = AGEGR1N)

> head(adsl_ct, n=10)

Author: Selection applied within `head()`

```

> 

Author: TOC added in `adsl.qmd` and `advs.qmd`

> 

Author: option 1 and option 2 swiped

)

adsl_cat <- derive_vars_cat(
   dataset = adsl_preds,
   definition = agegr1_lookup
 )

head(adsl_cat %>%  select(USUBJID, AGE, AGEU, AGEGR1, AGEGR1N), n=10)
```

Option 3: We can also solve this subgroups task with custom functions. 

```{r grouping_option_3}
format_agegr1 <- function(var_input) {
  case_when(
    var_input < 18 ~ "<18",
    between(var_input, 18, 64) ~ "18-64",
    var_input > 64 ~ ">64",
    TRUE ~ "Missing"
  )
}

format_agegr1n <- function(var_input) {
  case_when(
    var_input < 18 ~ 1,
    between(var_input, 18, 64) ~ 2,
    var_input > 64 ~ 3,
    TRUE ~ 4
  )
}

adsl_cust <- adsl_preds %>%
  mutate(
    AGEGR1 = format_agegr1(AGE),
    AGEGR1N = format_agegr1n(AGE)
    )

head(adsl_cust %>%  select(USUBJID, AGE, AGEU, AGEGR1, AGEGR1N), n=10)
```

Using a similar philosophy we can create the numeric version of `RACE` using the controlled terminology stored in the `{metacore}` object with the `metatools::create_var_from_codelist()` function.

```{r codelist}
adsl_ct <- adsl_ct %>%
   create_var_from_codelist(metacore = metacore,
                            input_var = RACE,
                            out_var = RACEN)

> head(adsl_ct, n=10)

Author: Selection applied within `head()`

```

### Exposure derivations

Now we have sorted out what we can easily do with controlled terminology it is time to start deriving some variables.
Here you could refer directly to using the `{admiral}` template and [vignette](https://pharmaverse.github.io/admiral/articles/adsl.html) in practice,
but for the purpose of this end-to-end ADaM vignette we will share a few exposure derivations from there.
We derive the start and end of treatment (which requires dates to first be converted from DTC to DTM), the treatment start time, the treatment duration, and the safety population flag.

> Note that the populations flags are mainly company/study specific no dedicated functions are provided, but in most cases they can easily be derived using `admiral::derive_var_merged_exist_flag()`.

Author: Clarified adding ", therefore,"

Now we have the base dataset, we can start to create some variables.
There are a few options to create grouping variables and their corresponding numeric variables.

**Option 1:** We can start with creating the subgroups using the **`admiral::derive_vars_cat()`** function available since `{admiral}` v1.2.0. 
This function is especially useful if more than one variable needs to be created for each condition, e.g., `AGEGR1` and `AGEGR1N`.
Additionally, one needs to be careful when considering the order of the conditions in the lookup table.
The category is assigned based on the first match. That means *catch-all* conditions must come after specific conditions, e.g. `!is.na(AGE)` must come after `between(AGE, 18, 64)`.

```{r grouping_option_1}
agegr1_lookup <- exprs(
  ~condition,            ~AGEGR1, ~AGEGR1N,
  is.na(AGE),          "Missing",        4,
  AGE < 18,                "<18",        1,
  between(AGE, 18, 64),  "18-64",        2,
  !is.na(AGE),             ">64",        3
)

adsl_cat <- derive_vars_cat(
   dataset = adsl_preds,
   definition = agegr1_lookup
 )

head(adsl_cat %>% select(STUDYID, USUBJID, AGE, AGEU, AGEGR1, AGEGR1N), n=10)
```

> Option 2: We can also create the subgroups using the controlled terminology, in this case `AGEGR1`.

Reviewer 2: ```suggestion
**Option 2:** We can also create the subgroups using the controlled terminology, in this case `AGEGR1`.
```

The metacore object holds all the metadata needed to make `ADSL`. Part of that metadata is the controlled terminology,
which can help automate the creation of subgroups. We can look into the `{metacore}` object and see the controlled terminology for `AGEGR1`.

```{r}
get_control_term(metacore, variable = AGEGR1)
```

Because this controlled terminology is written in a fairly standard format we can automate the creation of `AGEGR1`.
The function `metatools::create_cat_var()` takes in a `{metacore}` object, a reference variable -
in this case `AGE` because that is the continuous variable `AGEGR1` is created from, and the name of the sub-grouped variable.
It will take the controlled terminology from the sub-grouped variable and group the reference variables accordingly.

```{r grouping_option_2}
adsl_ct <- adsl_preds %>%
   create_cat_var(metacore, ref_var = AGE,
                  grp_var = AGEGR1, num_grp_var = AGEGR1N)

head(adsl_ct %>% select(STUDYID, USUBJID, AGE, AGEU, AGEGR1, AGEGR1N), n=10)
```

> Option 3: Another option to solve this subgroups task is to use custom functions. 

Reviewer 2: ```suggestion
**Option 3:** Another option to solve this subgroups task is to use custom functions. 
```

  # Treatment Start Datetime
  derive_vars_merged(
    dataset_add = ex_ext,
    filter_add = (EXDOSE > 0 |
      (EXDOSE == 0 &
        str_detect(EXTRT, "PLACEBO"))) & !is.na(EXSTDTM),
    new_vars = exprs(TRTSDTM = EXSTDTM, TRTSTMF = EXSTTMF),
    order = exprs(EXSTDTM, EXSEQ),
    mode = "first",
    by_vars = exprs(STUDYID, USUBJID)
  ) %>%
  # Treatment End Datetime
  derive_vars_merged(
    dataset_add = ex_ext,
    filter_add = (EXDOSE > 0 |
      (EXDOSE == 0 &
        str_detect(EXTRT, "PLACEBO"))) & !is.na(EXENDTM),
    new_vars = exprs(TRTEDTM = EXENDTM, TRTETMF = EXENTMF),
    order = exprs(EXENDTM, EXSEQ),
    mode = "last",
    by_vars = exprs(STUDYID, USUBJID)
  ) %>%
  # Treatment Start and End Date
   derive_vars_dtm_to_dt(source_vars = exprs(TRTSDTM, TRTEDTM)) %>%  # Convert Datetime variables to date
  # Treatment Start Time
  derive_vars_dtm_to_tm(source_vars = exprs(TRTSDTM)) %>% 
  # Treatment Duration
   derive_var_trtdurd() %>%
  # Safety Population Flag
   derive_var_merged_exist_flag(
     dataset_add = ex,
     by_vars = exprs(STUDYID, USUBJID),
     new_var = SAFFL,
     condition = (EXDOSE > 0 | (EXDOSE == 0 & str_detect(EXTRT, "PLACEBO")))
   ) %>%

>    drop_unspec_vars(metacore) # This will drop any columns that are not specified in the metacore object

Author: dropped

        str_detect(EXTRT, "PLACEBO"))) & !is.na(EXSTDTM),
    new_vars = exprs(TRTSDTM = EXSTDTM, TRTSTMF = EXSTTMF),
    order = exprs(EXSTDTM, EXSEQ),
    mode = "first",
    by_vars = exprs(STUDYID, USUBJID)
  ) %>%
  # Treatment End Datetime
  derive_vars_merged(
    dataset_add = ex_ext,
    filter_add = (EXDOSE > 0 |
      (EXDOSE == 0 &
        str_detect(EXTRT, "PLACEBO"))) & !is.na(EXENDTM),
    new_vars = exprs(TRTEDTM = EXENDTM, TRTETMF = EXENTMF),
    order = exprs(EXENDTM, EXSEQ),
    mode = "last",
    by_vars = exprs(STUDYID, USUBJID)
  ) %>%
  # Treatment Start and End Date
  derive_vars_dtm_to_dt(source_vars = exprs(TRTSDTM, TRTEDTM)) %>%  # Convert Datetime variables to date
  # Treatment Start Time
  derive_vars_dtm_to_tm(source_vars = exprs(TRTSDTM)) %>% 
  # Treatment Duration
  derive_var_trtdurd() %>%
  # Safety Population Flag
  derive_var_merged_exist_flag(
     dataset_add = ex,
     by_vars = exprs(STUDYID, USUBJID),
     new_var = SAFFL,
     condition = (EXDOSE > 0 | (EXDOSE == 0 & str_detect(EXTRT, "PLACEBO")))
   )

head(adsl_raw %>% select(STUDYID, USUBJID, TRTSDTM, TRTSTM, TRTSTMF, TRTSDT, TRTEDTM, TRTETMF, TRTEDT, TRTDURD, SAFFL), n=10)
```

> This call returns the original data frame with the column `TRTSDTM`, `TRTSTMF`, `TRTEDTM`, and `TRTETMF` added.

Author: Rephrased
